### PR TITLE
Add quickcheck

### PR DESCRIPTION
**Pull request by f-f**
This PR adds quickcheck - a few notes:
- Having `mod` for `Number` implemented as `flmod0` breaks quickcheck (or more specifically the part where it integrates with [LCG](https://github.com/purescript/purescript-lcg). Switching to `flmod` fixes the issue.
- I updated Prelude because I needed to debug the above with the JS backend
- The FFI for `Gen` is doing `foreign-alloc` and `foreign-free` because that was the only way I found that would do a cast. @anttih do you have a better idea?

### CODE REVIEW

**Comment by f-f** [comment]
Ha, more `mod` fun. So, the original issue was that our remainder operation `flmod0` was returning `-1.0` for an odd number:
```
> (flmod0 1678280651.0 2.0)
-1.0
> 
```

and this didn't match the behavior in node:
```
Welcome to Node.js v18.18.2.
Type ".help" for more information.
> (1678280651.0 % 2.0)
1
```

Now, `flmod` matches node on that, but negative numbers are then treated in a different way, and `flmod0` matches node, but the new `flmod` doesn't:
```
> (flmod0 -12.0 5.0)
-2.0
> (flmod -12.0 5.0)
3.0
```

and in in node:
```
> (-12.0) % 5.0
-2
```

**Comment by anttih** [comment]
>     * The FFI for `Gen` is doing `foreign-alloc` and `foreign-free` because that was the only way I found that would do a cast. @anttih do you have a better idea?

First, does `float32ToInt32` need to return a 32-bit fixnum given that `purescm` `Int` is not necessarily 32 bits in length (it is just whatever fixnum is)?

Anyways I think you could just do `(foreign-ref 'fixnum (object->reference-address n) 0)` to cast a flonum to fixnum. But TBH I don't know if that works here in the way we want it. It certainly is not "float 32 to int 32" anymore, but it does a similar cast from flonum to fixnum.

**Comment by f-f** [comment]
> First, does `float32ToInt32` need to return a 32-bit fixnum

I think so, the LCG implementation seem to rely on seeds being int32 - it's using a [Mersenne prime at the boundary of ints](https://github.com/purescript/purescript-lcg/blob/67c6c6483a563a59ae036d9dca0f1be2835326a5/src/Random/LCG.purs#L72).

So if we can do `(foreign-ref 'integer32 (object->reference-address n) 0)` we're good I think?
The neat thing about the current implementation is that it perfectly matches the JS one.

**Comment by f-f** [comment]
Yeah, the method above works but returns different results from JS:
```
> (define n 3.0)
> (foreign-ref 'integer-32 (object->reference-address n) 0)
134217728
```

vs this in JS:
```
> float32ToInt32(3.0)
1077936128
```

**Comment by f-f** [comment]
@anttih this is good to go now!

File: prelude/src/Data/Show.js [hunk 1]
```diff
@@ -59,15 +57,3 @@ exports.showArrayImpl = function (f) {
...
-  return function (tail) {
-    return [head].concat(tail);
-  };
-};
-
-exports.join = function (separator) {
-  return function (xs) {
-    return xs.join(separator);
-  };
-};
```
Comment (anttih) [inline prelude/src/Data/Show.js:73]: Maybe these scheme definitions could also be removed?

File: prelude/src/Data/Reflectable.js [hunk 2]
```diff
@@ -0,0 +1,5 @@
+// module Data.Reflectable
+
+export const unsafeCoerce = function (arg) {
```
Comment (anttih) [inline prelude/src/Data/Reflectable.js:3]: Do we need this definition for scheme?

File: quickcheck/src/Test/QuickCheck/Gen.ss [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+  (export float32ToInt32
+  )
+
+  (import (only (rnrs base) define let lambda quote)
+          (prefix (chezscheme) scm:)
+  )
+
+  (define float32ToInt32
+    (lambda (f)
+      (let ([buf (scm:foreign-alloc 4)])
```
Comment (anttih) [inline quickcheck/src/Test/QuickCheck/Gen.ss]:
```suggestion
      (let ([buf (scm:foreign-alloc (scm:foreign-sizeof 'single-float)])
```

**Comment by anttih** [comment]
> So if we can do `(foreign-ref 'integer32 (object->reference-address n) 0)` we're good I think? The neat thing about the current implementation is that it perfectly matches the JS one.

`object->reference-address` doesn't seem to work actually. I'm getting different results based on the expression I pass to it. I think what you have is fine.

File: prelude/src/Data/Reflectable.js (hunk 2 above)
Comment (f-f) [reply prelude/src/Data/Reflectable.js:3]: Ah yes, I typed it down and forgot to commit it

File: quickcheck/src/Test/QuickCheck/Gen.ss [hunk 4]
```diff
@@ -10,7 +10,7 @@
 
   (define float32ToInt32
     (lambda (f)
-      (let ([buf (scm:foreign-alloc 4)])
```
Comment (anttih) [inline quickcheck/src/Test/QuickCheck/Gen.ss]: Oops sorry I think I had a syntax error here. A missing closing paren.

Comment (f-f) [reply quickcheck/src/Test/QuickCheck/Gen.ss]: Tests pass though?? 🤔

Comment (f-f) [reply quickcheck/src/Test/QuickCheck/Gen.ss]: Well, I tried rerunning CI and it doesn't seem to pick up the wrong changes 🤷

**Review by anttih — APPROVED** [review]

**Pull request merged**